modules/porous_flow/src/materials/PorousFlowMaterialPermeabilityFromPorosity.C

```diff
@@ -0,0 +1,82 @@
+  params.addRequiredParam<UserObjectName>("PorousFlowDictator_UO", "The UserObject that holds the list of Porous-Flow variable names.");
+  params.addClassDescription("This Material calculates the permeability tensor from a specified function of porosity");
+  return params;
+}
+
+PorousFlowMaterialPermeabilityFromPorosity::PorousFlowMaterialPermeabilityFromPorosity(const InputParameters & parameters) :
+    DerivativeMaterialInterface<Material>(parameters),
+    _k0_set( parameters.isParamValid("k0") ),
+    _phi0_set( parameters.isParamValid("phi0") ),
+    _f_set( parameters.isParamValid("f") ),
+    _d_set( parameters.isParamValid("d") ),
+    _k_anisotropy_set( parameters.isParamValid("k_anisotropy") ),
+    _k0( _k0_set ? getParam<Real>("k0") : -1 ),
+    _phi0( _phi0_set ? getParam<Real>("phi0") : -1 ),
+    _f( _f_set ? getParam<Real>("f") : -1 ),
+    _d( _d_set ? getParam<Real>("d") : -1 ),
+    _m(getParam<Real>("m")),
+    _n(getParam<Real>("n")),
+    _k_anisotropy( _k_anisotropy_set ? getParam<RealTensorValue>("k_anisotropy") : getParam<RealTensorValue>("1 0 0  0 1 0  0 0 1")),
+    _porosity_qp(getMaterialProperty<Real>("PorousFlow_porosity_qp")),
+    _dporosity_qp_dvar(getMaterialProperty<std::vector<Real> >("dPorousFlow_porosity_qp_dvar")),
+    _poroperm_function(getParam<MooseEnum>("poroperm_function")),
+    _PorousFlow_name_UO(getUserObject<PorousFlowDictator>("PorousFlowDictator_UO")),
+    _num_var(_PorousFlow_name_UO.numVariables()),
+    _permeability(declareProperty<RealTensorValue>("PorousFlow_permeability")),
+    _dpermeability_dvar(declareProperty<std::vector<RealTensorValue> >("dPorousFlow_permeability_dvar"))
+{
+  switch (_poroperm_function)
+  {
+    case 0: // kozeny_carman_fd2
+      if (!(_f_set && _d_set))
+        mooseError("You must specify f and d in order to use kozeny_carman_fd2 in PorousFlowMaterialPermeabilityFromPorosity");
+      _mult = _f*_d*_d;
+      break;
+    case 1: // kozeny_carman_phi0
+      if (!(_k0_set && _phi0_set))
+        mooseError("You must specify k0 and phi0 in order to use kozeny_carman_phi0 in PorousFlowMaterialPermeabilityFromPorosity");
+      _mult = _k0*pow(1.0-_phi0,_m)/pow(_phi0,_n);
+      break;
+  }
+}
+
+void
+PorousFlowMaterialPermeabilityFromPorosity::initQpStatefulProperties()
+{
+  // **********  Not sure if this is the right thing or even possible! If not need to figure out an alternative for initialising perm.
+  _permeability[_qp] = _k_anisotropy*_mult*pow(_porosity_qp[_qp],_n)/pow(1.0-_porosity_qp[_qp],_m);
+}
+
+void
+PorousFlowMaterialPermeabilityFromPorosity::computeQpProperties()
+{
+  _permeability[_qp] = _k_anisotropy*_mult*pow(_porosity_qp[_qp],_n)/pow(1.0-_porosity_qp[_qp],_m);
+
+  _dpermeability_dvar[_qp].resize(_num_var, RealTensorValue());
+  for (unsigned v = 0; v < _num_var; ++v)
+    _dpermeability_dvar[_qp][v] = _dporosity_qp_dvar[_qp][v]*_permeability[_qp] * (_n/_porosity_qp[_qp] + _m/(1.0-_porosity_qp[_qp]));
```

Comment:
Would the test need to have porosity changing with time, otherwise there would be no point?

From: Andy Wilkins [mailto:notifications@github.com]
Sent: Monday, 1 August 2016 1:45 PM
To: WilkAndy/moose <moose@noreply.github.com>
Cc: Sheldon, Heather (Mineral Resources, Kensington WA) <Heather.Sheldon@csiro.au>; Author <author@noreply.github.com>
Subject: Re: [WilkAndy/moose] Initial porous flow 6845 - poroperm functions (#45)


In modules/porous_flow/src/materials/PorousFlowMaterialPermeabilityFromPorosity.C<https://github.com/WilkAndy/moose/pull/45#discussion_r72926260>:

> +

> +void

> +PorousFlowMaterialPermeabilityFromPorosity::initQpStatefulProperties()

> +{

> +  // **********  Not sure if this is the right thing or even possible! If not need to figure out an alternative for initialising perm.

> +  _permeability[_qp] = _k_anisotropy*_mult*pow(_porosity_qp[_qp],_n)/pow(1.0-_porosity_qp[_qp],_m);

> +}

> +

> +void

> +PorousFlowMaterialPermeabilityFromPorosity::computeQpProperties()

> +{

> +  _permeability[_qp] = _k_anisotropy*_mult*pow(_porosity_qp[_qp],_n)/pow(1.0-_porosity_qp[_qp],_m);

> +

> +  _dpermeability_dvar[_qp].resize(_num_var, RealTensorValue());

> +  for (unsigned v = 0; v < _num_var; ++v)

> +    _dpermeability_dvar[_qp][v] = _dporosity_qp_dvar[_qp][v]*_permeability[_qp] * (_n/_porosity_qp[_qp] + _m/(1.0-_porosity_qp[_qp]));

This derivative looks correct. Please could you check the MOOSE Jacobian using the snes_type=test using the PorousFlowAdvectiveFlux. Look at the tests/jacobian/fflux*.i to copy. The reason i'm asking is that Kernel contains dpermeability_dvar, but until now it has always been zero, so the Kernel Jacobian itself might actually be wrong, even if your derivative is correct (or the Kernel Jacobian might be right and your derivative wrong, or both wrong, or both right!)

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/WilkAndy/moose/pull/45/files/70988634d2f96dd7869d443ace65039c2244f23d#r72926260>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AKJ_8hGIJziUgP-uplqGHJpI5aSur8Reks5qbYfNgaJpZM4JZPBg>.